File: examples/projects/benchmarks/dyn_dispatch/dyn_dispatch.go
// 基准: dyn_dispatch —— 2000 万次接口方法多态分派
// 与 dyn_dispatch.rl 逻辑严格一致。输出 = 70000000
// 实现: Go interface（itab 间接调用），与 Rlyeh dyn Trait 胖指针 + vtable 分派对称
package main

import "fmt"

type Shape interface {
	sides() int64
}

type Tri struct{ n int64 }
type Quad struct{ n int64 }

func (t *Tri) sides() int64  { return t.n + 3 }
func (q *Quad) sides() int64 { return q.n + 4 }

func main() {
	var d1 Shape = &Tri{0}
	var d2 Shape = &Quad{0}
	var sum int64
	for i := 0; i < 10000000; i++ {
		sum += d1.sides()
		sum += d2.sides()
	}
	fmt.Println(sum)
}

File: examples/projects/benchmarks/dyn_dispatch/dyn_dispatch.rs
// 基准: dyn_dispatch —— 2000 万次多态分派（trait 对象 vtable）
// 与 dyn_dispatch.rl 逻辑严格一致。输出 = 70000000
trait Shape {
    fn sides(&self) -> i64;
}

struct Tri {
    n: i64,
}
impl Shape for Tri {
    fn sides(&self) -> i64 { self.n + 3 }
}

struct Quad {
    n: i64,
}
impl Shape for Quad {
    fn sides(&self) -> i64 { self.n + 4 }
}

fn main() {
    let t = Tri { n: 0 };
    let q = Quad { n: 0 };
    let d1: &dyn Shape = &t;
    let d2: &dyn Shape = &q;
    let mut sum: i64 = 0;
    for _ in 0..10_000_000 {
        sum += d1.sides();
        sum += d2.sides();
    }
    println!("{}", sum);
}
